python-package/xgboost/spark/core.py

```diff
@@ -0,0 +1,834 @@
+
+
+class _SparkXGBEstimator(Estimator, _XgboostParams, MLReadable, MLWritable):
+    def __init__(self):
+        super().__init__()
+        self._set_xgb_params_default()
+        self._set_fit_params_default()
+        self._set_predict_params_default()
+        # Note: The default value for arbitraryParamsDict must always be empty dict.
+        #  For additional settings added into "arbitraryParamsDict" by default,
+        #  they are added in `setParams`.
+        self._setDefault(
+            num_workers=1,
+            use_gpu=False,
+            force_repartition=False,
+            feature_names=None,
+            feature_types=None,
+            arbitraryParamsDict={}
+        )
+
+    def setParams(self, **kwargs):
+        _extra_params = {}
+        if 'arbitraryParamsDict' in kwargs:
+            raise ValueError("Wrong param name: 'arbitraryParamsDict'.")
+
+        for k, v in kwargs.items():
+            if k in _pyspark_param_alias_map:
+                real_k = _pyspark_param_alias_map[k]
+                if real_k in kwargs:
+                    raise ValueError(f"You should set only one of param '{k}' and '{real_k}'")
+                k = real_k
+
+            if self.hasParam(k):
+                self._set(**{str(k): v})
+            else:
+                if k in _unsupported_xgb_params or \
+                        k in _unsupported_fit_params or \
+                        k in _unsupported_predict_params:
+                    raise ValueError(f"Unsupported param '{k}'.")
+                _extra_params[k] = v
+        _existing_extra_params = self.getOrDefault(self.arbitraryParamsDict)
+        self._set(arbitraryParamsDict={**_existing_extra_params, **_extra_params})
+
+    @classmethod
+    def _pyspark_model_cls(cls):
+        """
+        Subclasses should override this method and
+        returns a _XgboostModel subclass
+        """
+        raise NotImplementedError()
+
+    def _create_pyspark_model(self, xgb_model):
+        return self._pyspark_model_cls()(xgb_model)
+
+    @classmethod
+    def _convert_to_classifier(cls, booster):
+        clf = XGBClassifier()
+        clf._Booster = booster
+        return clf
+
+    @classmethod
+    def _convert_to_regressor(cls, booster):
+        reg = XGBRegressor()
+        reg._Booster = booster
+        return reg
+
+    def _convert_to_model(self, booster):
+        if self._xgb_cls() == XGBRegressor:
+            return self._convert_to_regressor(booster)
+        elif self._xgb_cls() == XGBClassifier:
+            return self._convert_to_classifier(booster)
+        else:
+            return None  # check if this else statement is needed.
+
+    def _query_plan_contains_valid_repartition(self, dataset):
+        """
+        Returns true if the latest element in the logical plan is a valid repartition
+        """
+        num_partitions = dataset.rdd.getNumPartitions()
+        query_plan = dataset._sc._jvm.PythonSQLUtils.explainString(
```

Comment:
Could you please add some comments on how it works? Seems you are reaching to internal API and it would be great to have a brief note.